### PR TITLE
Ignore private/sealed abstract class/traits

### LINKED ISSUE #14579: forward port knowledge of private and sealed traits in exhaustivity warnings
port fixes from scala 2 that closed https://github.com/scala/scala/pull/9211

## Compiler version

3.1.1

## Minimized code

```Scala
trait A {
  sealed abstract class X
  private class X1 extends X with X2 { }
  private trait X2 extends X
  sealed trait X3 extends X

  def f(x: X) = x match {
    case _: X1 => 0
  }
}
```

## Output

```scala
-- [E029] Pattern Match Exhaustivity Warning: ----------------------------------
7 |  def f(x: X) = x match {
  |                ^
  |                match may not be exhaustive.
  |
  |                It would fail on pattern case: _: X2, _: X3
```

## Expectation

No warning, like scala 2

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes #14579

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 1]
```diff
@@ -628,7 +628,13 @@ class SpaceEngine(using Context) extends SpaceLogic {
       case tp if tp.classSymbol.isAllOf(JavaEnumTrait) =>
         tp.classSymbol.children.map(sym => Typ(sym.termRef, true))
       case tp =>
-        val children = tp.classSymbol.children
+        def getChildren(sym: Symbol): List[Symbol] =
+          sym.children.flatMap { child =>
+            if child eq sym then Nil // i3145: sealed trait Baz, val x = new Baz {}, Baz.children returns Baz...
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:633]: will there still show `_: Baz` to represent the anonymous cases?

**Comment by dwijnand** [comment]
lol

```scala
def map[F[_]](self: Tuple, f: [t] => t => F[t]): Tuple = self match {
  case EmptyTuple => self
  case _ => fromIArray(self.productIterator.map(f(_).asInstanceOf[Object]).toArray.asInstanceOf[IArray[Object]]) // TODO use toIArray
}
```

```
Error:  -- [E030] Match case Unreachable Error: /__w/dotty/dotty/library/src/scala/runtime/Tuples.scala:544:9 
Error:  544 |    case _ => fromIArray(self.productIterator.map(f(_).asInstanceOf[Object]).toArray.asInstanceOf[IArray[Object]]) // TODO use toIArray
Error:      |         ^
Error:      |         Unreachable case
Error:  one error found
```

Well the logic is right... outside of EmptyTuple, the Tuple class has no subclasses... 🙄   Do I special case tuple-vs-generic-tuple in SpaceEngine too??

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 2]
```diff
@@ -628,7 +628,14 @@ class SpaceEngine(using Context) extends SpaceLogic {
...
+            else if tp.classSymbol == defn.TupleClass || tp.classSymbol == defn.NonEmptyTupleClass then List(child) // Tuple is weird/different...
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:634]:
```suggestion
            else if tp.classSymbol == defn.TupleClass || tp.classSymbol == defn.NonEmptyTupleClass then List(child) // Tuple types are substituted for different classes at runtime
```

**Review by bishabosha — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala (hunk 2 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:634]: Actually, it makes me think: is there a reason TupleXXL doesn't extend Tuple and the TupleN classes made synthetically to extend Tuple?  Then we wouldn't need this.

**Pull request merged**